src/storage-client/src/source/persist_source.rs

```diff
@@ -202,65 +207,58 @@ where
+                                                    updates_session.give(&cap, (row, time, diff));
+                                                    updates += 1;
                                                 }
                                             }
                                             Err((err, time, diff)) => {
                                                 // Additional `until` filtering due to temporal filters.
                                                 if !until.less_equal(&time) {
-                                                    updates.push((Err(err), time, diff));
+                                                    err_session.give(&cap, (err, time, diff));
                                                 }
                                             }
                                         }
                                     }
                                 } else {
-                                    updates.push((Ok(row), time, diff));
+                                    updates_session.give(&cap, (row, time, diff));
+                                    updates += 1;
                                 }
                             }
                             (Ok(SourceData(Err(err))), Ok(())) => {
-                                updates.push((Err(err), time, diff));
+                                err_session.give(&cap, (err, time, diff));
                             }
                             // TODO(petrosagg): error handling
                             (Err(_), Ok(_)) | (Ok(_), Err(_)) | (Err(_), Err(_)) => {
                                 panic!("decoding failed")
                             }
                         }
                     }
-                    if yield_fn(decode_start, updates.len()) {
-                        // A large part of the point of yielding is to let later operators
-                        // reduce down the data, so emit what we have. Note that this means
-                        // we don't get to consolidate everything, but that's part of the
-                        // tradeoff in tuning yield_fn.
-                        differential_dataflow::consolidation::consolidate_updates(&mut updates);
-
-                        {
-                            // Do very fine-grained output activation/session
-                            // creation to ensure that we don't hold activated
-                            // outputs or sessions across await points, which
-                            // would prevent messages from being flushed from
-                            // the shared timely output buffer.
-                            let mut updates_output = updates_output.activate();
-                            updates_output.session(&cap).give_vec(&mut updates);
+                    if yield_fn(decode_start, updates) {
+                        updates = 0;
+                        unsafe {
```

Comment:
I try pretty hard to avoid unsafe whenever possible and would feel much more comfortable with merging this if we first did whatever refactor was necessary to avoid it. Maybe we could use Option instead and turn the unsafety into `.expect` panics?

Reply:
Addressing a shortcoming in Timely in https://github.com/TimelyDataflow/timely-dataflow/pull/496 to avoid the unsafe code here. I agree this is bad!